### PR TITLE
Split image metadata facade.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

This PR splits image metadata facade into 2 - one for client and one for the worker.
The worker-related facade remained the original name ImageMetadata and only contains functionality to retrieve and store metadata for published cloud images.
The new client facade, ImageMetadataManager, contains functionality for CRUD operations on image metadata. As these operations are currently only meant to be called as part of the experimental feature with commands sitting behind a feature flag, this facade is also put behind a feature flag.
There are some reservations around whether this facade should be behind a feature flag or not. The debate is around what is more important to us - usability or maintainability. From usability perspective, without the feature flag here, newer clients, that may have these commands exposed, will be able to use functionality without the need to upgrade controllers. However, to be more aligned with the approach on other experimental features, I have put the facade behind a feature flag for now.

The rest of the PR is moving tests and related code around into new locations as well as updating references and import statements.

Note that I have not actually changed the content of the code besides the basic move, rename and typo corrections.

## QA steps

There should be no changes to behavior - all unit and CI tests still pass.

## Documentation changes

n/a

## Bug reference

https://bugs.launchpad.net/juju/+bug/1703582

### CODE REVIEW

File: apiserver/facades/agent/imagemetadata/metadata.go [hunk 1]
```diff
@@ -4,7 +4,6 @@
 package imagemetadata
```
Comment (axw) [inline apiserver/facades/agent/imagemetadata/metadata.go]: the worker is only run by controllers, so it should go under facades/controller

Comment (axw) [inline apiserver/facades/agent/imagemetadata/metadata.go]: please make sure the imagemetadata facade allows only controllers to auth

File: apiserver/facades/client/imagemetadatamanager/metadata.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+// createAPI returns a new image metadata API facade.
+func createAPI(
+	st metadataAccess,
+	newEnviron func() (environs.Environ, error),
+	resources facade.Resources,
+	authorizer facade.Authorizer,
+) (*API, error) {
+	if !authorizer.AuthClient() && !authorizer.AuthController() {
```
Comment (axw) [inline apiserver/facades/client/imagemetadatamanager/metadata.go]: this facade should allow only clients

```go
if !authorizer.AuthClient() {
```

File: apiserver/facades/client/imagemetadatamanager/metadata.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+)
+
+var logger = loggo.GetLogger("juju.apiserver.imagemetadatamanager")
+
+// API is the concrete implementation of the api end point
+// for loud image metadata manipulations.
+type API struct {
+	metadata   metadataAccess
+	newEnviron func() (environs.Environ, error)
+	authorizer facade.Authorizer
```
Comment (axw) [inline apiserver/facades/client/imagemetadatamanager/metadata.go]: no need to store authorizer

File: apiserver/facades/client/imagemetadatamanager/metadata.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+		return stateenvirons.GetNewEnvironFunc(environs.New)(st)
+	}
+	return createAPI(getState(st), newEnviron, resources, authorizer)
+}
+
+// List returns all found cloud image metadata that satisfy
+// given filter.
+// Returned list contains metadata ordered by priority.
+func (api *API) List(filter params.ImageMetadataFilter) (params.ListCloudImageMetadataResult, error) {
+	if api.authorizer.AuthClient() {
```
Comment (axw) [inline apiserver/facades/client/imagemetadatamanager/metadata.go]: the whole facade should only allow clients now, so drop this check, and ones in other non-constructor methods

File: apiserver/facades/client/imagemetadatamanager/metadata.go (hunk 3 above)
Comment (anastasiamac) [inline apiserver/facades/client/imagemetadatamanager/metadata.go]: Yes, true now that they are split there is no need. It also means that we no longer need to have ControllerTag call - a lot cleaner code. Done.

File: apiserver/facades/agent/imagemetadata/metadata.go (hunk 1 above)
Comment (anastasiamac) [inline apiserver/facades/agent/imagemetadata/metadata.go]: done

File: apiserver/facades/client/imagemetadatamanager/metadata.go (hunk 2 above)
Comment (anastasiamac) [inline apiserver/facades/client/imagemetadatamanager/metadata.go]: done

File: apiserver/facades/client/imagemetadatamanager/metadata.go (hunk 4 above)
Comment (anastasiamac) [inline apiserver/facades/client/imagemetadatamanager/metadata.go]: done

**Comment by anastasiamac** [comment]
I'd rather add the tests as a follow-up as I wanted to address our current fake authoriser used for tests first - it's really something I wanted to do for a while and will make permission tests easier.

Also, it'd be great if these permission tests where done at a higher level now that we have split between client, controller, etc facades. In other words, something along the lines of "test that all facades in facades/client package are client scoped".

**Comment by anastasiamac** [comment]
$$merge$$

**Comment by axw** [comment]
> Also, it'd be great if these permission tests where done at a higher level now that we have split between client, controller, etc facades. In other words, something along the lines of "test that all facades in facades/client package are client scoped".

That would be great, if it's possible.

**Pull request merged**